File: graph/graph/src/main/groovy/net/twisterrob/gradle/graph/vis/VisualizerSettings.java
package net.twisterrob.gradle.graph.vis;

import java.io.*;
import java.nio.channels.OverlappingFileLockException;
import java.util.Properties;

import org.gradle.cache.PersistentCache;
import org.gradle.internal.Factory;

import com.sun.javafx.beans.annotations.NonNull;

public abstract class VisualizerSettings<Settings> implements Closeable {
	private final PersistentCache cache;

	protected VisualizerSettings(PersistentCache cache) {
		if (cache == null) {
			throw new NullPointerException("cache cannot be null");
		}
		this.cache = cache;
	}

	public Settings getSettings() {
		try {
			return cache.useCache("load properties", new Factory<Settings>() {
				@Override public Settings create() {
					File propsFile = new File(cache.getBaseDir(), getSettingsFileName());
					Properties props = new Properties();
					try {
						props.load(new FileReader(propsFile));
						return readSettings(props);
					} catch (FileNotFoundException ex) {
						return readSettings(new Properties()); // first startup
					} catch (IOException ex) {
						throw new IllegalStateException("Cannot read settings from " + propsFile, ex);
					}
				}
			});
		} catch (OverlappingFileLockException ex) {
			System.err.println("Cannot read settings, using defaults: " + ex);
			return createDefault();
		}
	}

	protected abstract Settings readSettings(@NonNull Properties props);

	public void setSettings(Settings settings) {
		try {
			final Properties props = writeSettings(settings);
			cache.useCache("save properties", new Runnable() {
				@Override public void run() {
					File propsFile = new File(cache.getBaseDir(), getSettingsFileName());
					try {
						props.store(new FileWriter(propsFile), null);
					} catch (IOException e) {
						throw new IllegalStateException("Cannot save settings to " + propsFile, e);
					}
				}
			});
		} catch (OverlappingFileLockException ex) {
			System.err.println("Cannot save settings: " + ex);
		}
	}
	protected abstract Properties writeSettings(Settings settings);

	protected abstract Settings createDefault();

	protected String getSettingsFileName() {
		return getClass().getName() + ".properties";
	}

	public void close() {
		cache.close();
	}
}

File: graph/src/main/kotlin/net/twisterrob/gradle/graph/vis/VisualizerSettings.kt
package net.twisterrob.gradle.graph.vis

import org.gradle.cache.PersistentCache
import java.io.Closeable
import java.io.File
import java.io.FileNotFoundException
import java.io.FileReader
import java.io.FileWriter
import java.io.IOException
import java.nio.channels.OverlappingFileLockException
import java.util.Properties

abstract class VisualizerSettings<Settings : Any> protected constructor(
	private val cache: PersistentCache
) : Closeable {

	protected val settingsFileName: String
		get() = this::class.java.name + ".properties"

	var settings: Settings
		get() = try {
			cache.useCache<Settings> {
				val propsFile = File(cache.baseDir, settingsFileName)
				val props = Properties()
				try {
					props.load(FileReader(propsFile))
					readSettings(props)
				} catch (ex: FileNotFoundException) {
					readSettings(Properties()) // first startup
				} catch (ex: IOException) {
					throw IllegalStateException("Cannot read settings from $propsFile", ex)
				}
			}
		} catch (ex: OverlappingFileLockException) {
			System.err.println("Cannot read settings, using defaults: $ex")
			createDefault()
		}
		set(settings) {
			try {
				val props = writeSettings(settings)
				cache.useCache {
					val propsFile = File(cache.baseDir, settingsFileName)
					try {
						props.store(FileWriter(propsFile), null)
					} catch (e: IOException) {
						throw IllegalStateException("Cannot save settings to $propsFile", e)
					}
				}
			} catch (ex: OverlappingFileLockException) {
				System.err.println("Cannot save settings: $ex")
			}
		}

	protected abstract fun readSettings(props: Properties): Settings
	protected abstract fun writeSettings(settings: Settings): Properties
	protected abstract fun createDefault(): Settings

	override fun close() {
		cache.close()
	}
}
